### PR TITLE
[mlir] Expose MLIR_CUDA_CONVERSIONS_ENABLED in mlir-config.h.

### PR DESCRIPTION
**Pull request by ingomueller-net**
~~The `BUILD` files of two CUDA-related build targets were missing that definition such that the corresponding MLIR target libraries did not actually work.~~

That macro was not defined in some cases and thus yielded warnings if compiled with `-Wundef`. In particular, they were not defined in the BUILD files, so the GPU targets were broken when built with Bazel. This commit exposes mentioned CMake variable through mlir-config.h and uses the macro that is introduced with the same name. This replaces the macro MLIR_CUDA_CONVERSIONS_ENABLED, which the CMake files previously defined manually.

### CODE REVIEW

**Review by ftynse — APPROVED** [review]
As an FYI, the code relying on this does specifically: [`#if MLIR_CUDA_CONVERSIONS_ENABLED == 1`](https://github.com/llvm/llvm-project/blob/f2bb6c4415954535b32780cd5fb48411ebe0042c/mlir/lib/Target/LLVM/NVVM/Target.cpp#L159) so you may want to define the expected value here, or update that code to use an `#ifdef` instead.

**Comment by ingomueller-net** [comment]
> As an FYI, the code relying on this does specifically: [`#if MLIR_CUDA_CONVERSIONS_ENABLED == 1`](https://github.com/llvm/llvm-project/blob/f2bb6c4415954535b32780cd5fb48411ebe0042c/mlir/lib/Target/LLVM/NVVM/Target.cpp#L159) so you may want to define the expected value here, or update that code to use an `#ifdef` instead.

@joker-eph raised [arguments](https://github.com/llvm/llvm-project/pull/82988#pullrequestreview-1902317048) against using `#ifdef` in #82988. I have thus rewritten this commit to use only `#if X`.

Note that this includes changes from `#if X == 1` to `#if X` because I don't see any argument in favor of the former but the discussion in the other PR was still open, so I am still happy to revert that part.

I have compiled with OSS Bazel with `-Wundef` and don't get any warnings for this variable anymore. Also, I have compiled with CMake successfully. I still need to test the current form against our internal code base and will report back as soon as I have results.

**Comment by ingomueller-net** [comment]
Also, there are a few variables that have the same issue. Once we are happy with this approach, I might submit PR to apply it to them as well. Once that is done, maybe we can think about adding `-Wundef` to MLIR targets?

File: utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel [hunk 1]
```diff
@@ -552,7 +552,7 @@ cc_library(
 cc_library(
     name = "TestTransforms",
     srcs = glob(["lib/Transforms/*.cpp"]),
-    defines = ["MLIR_CUDA_CONVERSIONS_ENABLED"],
+    defines = ["MLIR_ENABLE_CUDA_CONVERSIONS"],
```
Comment (joker-eph) [inline utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel:555]: There is a discrepancy between the change in Cmake and Bazel: shouldn't remove this and  rely on mlir-config.h  instead?

**Review by joker-eph — APPROVED** [review]
LG, other than the Bazel change I'm unsure about

File: utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel (hunk 1 above)
Comment (ingomueller-net) [reply utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel:555]: I need to check this tomorrow at work but my guess was that this is activated by force here for testing. Not 100% sure now if that (1) is true and (2) makes sense.

Comment (joker-eph) [reply utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel:555]: No I think until now this matching CMake: it was defined on the command line. But now that we move it to the config.h, you removed it from the CMake definitions so I would think we should match this in bazel

Comment (ingomueller-net) [reply utils/bazel/llvm-project-overlay/mlir/test/BUILD.bazel]: Yeah, I think you are right. The last commit makes that change.

**Review by ingomueller-net — COMMENTED** [review]
OK, issue addressed. This version also passes all our internal tests.

**Pull request merged**